### PR TITLE
Added functionality to weight Cartesian points in Pca fits.

### PR DESCRIPTION
**Pull request by StevenGreen1**

### CODE REVIEW

File: larpandoracontent/LArHelpers/LArPcaHelper.cc [hunk 1]
```diff
@@ -18,17 +18,29 @@ namespace lar_content
 
 void LArPcaHelper::RunPca(const CaloHitList &caloHitList, CartesianVector &centroid, EigenValues &outputEigenValues, EigenVectors &outputEigenVectors)
 {
-    CartesianPointVector pointVector;
+    CartesianPointDoublePairVector cartesianPointDoublePairVector;
 
     for (const CaloHit *const pCaloHit : caloHitList)
-        pointVector.push_back(pCaloHit->GetPositionVector());
+        cartesianPointDoublePairVector.push_back(std::make_pair(pCaloHit->GetPositionVector(), 1.0));
```
Comment (PandoraPFA) [inline larpandoracontent/LArHelpers/LArPcaHelper.cc]: As discussed, create LArObjectHelper with just the existing LArPfoHelper::TypeAdaptor in initially. Obviously, then remove this from LArPfoHelper.

Comment (StevenGreen1) [reply larpandoracontent/LArHelpers/LArPcaHelper.cc]: Done and testing pca now.

**Comment by StevenGreen1** [comment]
Pca fit was noted to not use the weights when calculating the centroid position.  This has been corrected and tested and now works.  Furthermore the RunPca has been templated to work using a CaloHitList and a CartesianPointVector.

File: larpandoracontent/LArHelpers/LArObjectHelper.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    public:
+        /**
+         *  @brief  Get the associated position
+         *
+         *  @param  t the input object
+         *
+         *  @return the associated position
+         */
+        template<typename T>
+        static const pandora::CartesianVector GetPosition(const T &t);
```
Comment (PandoraPFA) [inline larpandoracontent/LArHelpers/LArObjectHelper.h]: No need for const by value return type (as it was previously!)

File: larpandoracontent/LArHelpers/LArPcaHelper.h [hunk 3]
```diff
@@ -20,27 +20,28 @@ class LArPcaHelper
 public:
     typedef pandora::CartesianVector EigenValues;
     typedef std::vector<pandora::CartesianVector> EigenVectors;
+    typedef std::vector<std::pair<const pandora::CartesianVector, double> > CartesianPointDoublePairVector;
```
Comment (PandoraPFA) [inline larpandoracontent/LArHelpers/LArPcaHelper.h]: Suggest intermediate typedef for pair.

File: larpandoracontent/LArHelpers/LArPcaHelper.cc [hunk 4]
```diff
@@ -16,19 +17,20 @@ using namespace pandora;
...
-void LArPcaHelper::RunPca(const CaloHitList &caloHitList, CartesianVector &centroid, EigenValues &outputEigenValues, EigenVectors &outputEigenVectors)
+template <typename T>
+void LArPcaHelper::RunPca(const T &t, CartesianVector &centroid, EigenValues &outputEigenValues, EigenVectors &outputEigenVectors)
 {
-    CartesianPointVector pointVector;
+    CartesianPointDoublePairVector cartesianPointDoublePairVector;
 
-    for (const CaloHit *const pCaloHit : caloHitList)
-        pointVector.push_back(pCaloHit->GetPositionVector());
+    for (const auto &iter : t)
```
Comment (PandoraPFA) [inline larpandoracontent/LArHelpers/LArPcaHelper.cc]: Alternative name to iter? Maybe e.g. positionValue or similar. Previously point.

File: larpandoracontent/LArHelpers/LArPcaHelper.cc [hunk 5]
```diff
@@ -44,18 +46,21 @@ void LArPcaHelper::RunPca(const CartesianPointVector &pointVector, CartesianVect
...
+        meanPosition[1] += point.GetY() * weight;
+        meanPosition[2] += point.GetZ() * weight;
+        nThreeDWeightedHitsAsDouble += weight;
     }
 
-    const double nThreeDHitsAsDouble(static_cast<double>(pointVector.size()));
-    meanPosition[0] /= nThreeDHitsAsDouble;
-    meanPosition[1] /= nThreeDHitsAsDouble;
-    meanPosition[2] /= nThreeDHitsAsDouble;
+    meanPosition[0] /= nThreeDWeightedHitsAsDouble;
```
Comment (PandoraPFA) [inline larpandoracontent/LArHelpers/LArPcaHelper.cc]: Appropriate action if `(std::fabs(nThreeDWeightedHitsAsDouble) < std::numeric_limits<double>::epsilon())`

File: larpandoracontent/LArHelpers/LArPcaHelper.cc [hunk 6]
```diff
@@ -44,18 +46,21 @@ void LArPcaHelper::RunPca(const CartesianPointVector &pointVector, CartesianVect
     }
 
     double meanPosition[3] = {0., 0., 0.};
```
Comment (PandoraPFA) [inline larpandoracontent/LArHelpers/LArPcaHelper.cc]: Could upgrade, with suitable tests to check unchanged, as discussed.

File: larpandoracontent/LArHelpers/LArPcaHelper.cc [hunk 7]
```diff
@@ -110,7 +116,7 @@ void LArPcaHelper::RunPca(const CartesianPointVector &pointVector, CartesianVect
     typedef std::vector<EigenValColPair> EigenValColVector;
 
     EigenValColVector eigenValColVector;
-    const auto &resultEigenMat(eigenMat.eigenvalues());
+    const auto &resultEigenMat(eigenMat.eigenvalues());//------------------------------------------------------------------------------------------------------------------------------------------
```
Comment (PandoraPFA) [inline larpandoracontent/LArHelpers/LArPcaHelper.cc]: Cosmetic.

File: larpandoracontent/LArHelpers/LArPcaHelper.h [hunk 8]
```diff
@@ -20,27 +20,28 @@ class LArPcaHelper
...
+     *  @brief  Run principal component analysis using weighted input Cartesian vectors (TPC_VIEW_U,V,W or TPC_3D; all treated as 3D points)
      *
-     *  @param  pointVector the vector of positions
+     *  @param  pointVector a vector of pairs of positions and weights 
      *  @param  centroid to receive the centroid position
      *  @param  outputEigenValues to receive the eigen values
      *  @param  outputEigenVectors to receive the eigen vectors
      */
-    static void RunPca(const pandora::CartesianPointVector &pointVector, pandora::CartesianVector &centroid, EigenValues &outputEigenValues,
+    static void RunPca(const CartesianPointDoublePairVector &pointVector, pandora::CartesianVector &centroid, EigenValues &outputEigenValues,
```
Comment (PandoraPFA) [inline larpandoracontent/LArHelpers/LArPcaHelper.h]: Whitespace(!)

File: larpandoracontent/LArHelpers/LArObjectHelper.h (hunk 2 above)
Comment (StevenGreen1) [reply larpandoracontent/LArHelpers/LArObjectHelper.h]: Have made this correction.  To further help me understand this, am I right in assuming a const pointer return type (ala GetCaloHit in same class) is something that makes sense as the pointer being returned points to the same thing in both the TypeAdaptor class and where this function is called and that object is const?

**Pull request merged**